sound/soc/sof/ipc.c

```diff
@@ -398,17 +398,27 @@ static void ipc_msgs_rx(struct work_struct *work)
 	case SOF_IPC_FW_READY:
 		/* check for FW boot completion */
 		if (!sdev->boot_complete) {
-			if (sof_ops(sdev)->fw_ready)
-				err = sof_ops(sdev)->fw_ready(sdev, cmd);
+			/* check if fw_ready op is defined */
+			if (!sof_ops(sdev)->fw_ready) {
+				dev_err(sdev->dev,
+					"error: fw_ready op not defined\n");
+				goto out;
+			}
+
+			err = sof_ops(sdev)->fw_ready(sdev, cmd);
 			if (err < 0) {
-				dev_err(sdev->dev, "error: DSP firmware boot timeout %d\n",
+				/*
+				 * this indicates a mismatch in ABI
+				 * between the driver and fw
+				 */
+				dev_err(sdev->dev, "error: ABI mismatch %d\n",
 					err);
 			} else {
 				/* firmware boot completed OK */
 				sdev->boot_complete = true;
-				dev_dbg(sdev->dev, "booting DSP firmware completed\n");
-				wake_up(&sdev->boot_wait);
 			}
+out:
+			wake_up(&sdev->boot_wait);
```

Comment:
I really don't like the flow when fw_ready is not defined. We should test this in the core before enabling the IPC, not before using it in the RX.

Reply:
@plbossart fixed to make sure the madatory ops for ipc are defined before initializing ipc.